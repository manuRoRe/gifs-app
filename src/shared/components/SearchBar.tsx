import React, { useEffect, useState } from "react";

interface SearchBarProps {
  placeHolder?: string;
  onQuery: (query: string) => void;
}
export const SearchBar = ({
  placeHolder = "Buscar",
  onQuery,
}: SearchBarProps) => {
  const [query, setQuery] = useState("");

  useEffect(() => {
    const timeoutID = setTimeout(() => {
      onQuery(query);
    }, 700);

    return () => {
      clearTimeout(timeoutID);
    };
  }, [query, onQuery]);

  const handleSearch = () => {
    onQuery(query);
    /*  setQuery(""); */
  };

  const handleKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
    if (event.key === "Enter") {
      handleSearch();
    }
  };

  return (
    <div className="search-container">
      <input
        type="text"
        placeholder={placeHolder}
        value={query}
        onChange={(event) => setQuery(event.target.value)}
        onKeyDown={handleKeyDown}
      />
      <button type="submit" onClick={handleSearch}>
        Buscar
      </button>
    </div>
  );
};
