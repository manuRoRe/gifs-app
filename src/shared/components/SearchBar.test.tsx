import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { describe, expect, test, vi } from "vitest";
import { SearchBar } from "./SearchBar";

describe("SearchBar", () => {
  test("should render SearchBar", () => {
    const { container } = render(<SearchBar onQuery={() => {}} />);

    expect(container).toMatchSnapshot();
  });

  test("Should call onQuery with the correct value after 700ms", async () => {
    const onQuery = vi.fn();
    render(<SearchBar onQuery={onQuery} />);

    const input = screen.getByRole("textbox");
    fireEvent.change(input, { target: { value: "test" } });

    screen.debug();
    //! await new Promise((resolve) => setTimeout(resolve, 701));

    await waitFor(() => {
      expect(onQuery).toHaveBeenCalled();
      expect(onQuery).toHaveBeenCalledWith("test");
    });
  });
});
