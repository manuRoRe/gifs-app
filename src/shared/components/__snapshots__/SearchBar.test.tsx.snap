// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchBar > should render SearchBar 1`] = `
<div>
  <div
    class="search-container"
  >
    <input
      placeholder="Buscar"
      type="text"
      value=""
    />
    <button
      type="submit"
    >
      Buscar
    </button>
  </div>
</div>
`;
